core/vm/interpreter.go

```diff
@@ -918,3 +918,159 @@ func (in *GethEVMInterpreter) runPlain(state *InterpreterState, input []byte, re
 	}
 	return nil, nil
 }
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Abstracted interpreter with single step execution.
+
+// GethState represents the internal state of the interpreter.
+type GethState struct {
+	Contract *Contract // processed contract
+	Memory   *Memory   // bound memory
+	Stack    *Stack    // local stack
+	// For optimisation reason we're using uint64 as the program counter.
+	// It's theoretically possible to go above 2^64. The YP defines the PC
+	// to be uint256. Practically much less so feasible.
+	Pc          uint64 // program counter
+	Result      []byte // result of the opcode execution function
+	Err         error
+	CallContext *ScopeContext
+	Halted      bool
+
+	op OpCode // current opcode
+}
+
+func NewGethState(contract *Contract, memory *Memory, stack *Stack, Pc uint64) *GethState {
+	return &GethState{
+		Contract: contract,
+		Memory:   memory,
+		Stack:    stack,
+		Pc:       Pc,
+		CallContext: &ScopeContext{
+			Memory:   memory,
+			Stack:    stack,
+			Contract: contract,
+		},
+	}
+}
+
+func (in *GethEVMInterpreter) StartRun(state *GethState, input []byte, readOnly bool) {
+	// Increment the call depth which is restricted to 1024
+	in.evm.Depth++
+
+	// Make sure the readOnly is only set if we aren't in readOnly yet.
+	// This also makes sure that the readOnly flag isn't removed for child calls.
+	if readOnly && !in.readOnly {
+		in.readOnly = true
+	}
+
+	// Reset the previous call's return data. It's unimportant to preserve the old buffer
+	// as every returning call will return new data anyway.
+	in.returnData = nil
+
+	state.Contract.Input = input
+}
+
+func (in *GethEVMInterpreter) StopRun(readOnly bool) {
+	in.evm.Depth--
+
+	// Make sure the readOnly is only set if we aren't in readOnly yet.
+	// This also makes sure that the readOnly flag isn't removed for child calls.
+	if readOnly && in.readOnly {
+		in.readOnly = false
+	}
+}
```

Comment:
Are those needed?

Reply:
I do not think that we will need those for supporting recursive calls.

Reply:
Okay, I removed them and inlined the code into the `runPlain` function again.